WordPress/gutenberg
# Add new `registerInserterMediaCategory` API to make media categories extensible

## What?
In https://github.com/WordPress/gutenberg/pull/44918 and https://github.com/WordPress/gutenberg/issues/46222 we added a new media tab in the inserter and integrated Openverse respectively.  Currently the media categories are stored in a private block editor setting(`inserterMediaCategories`), which means they are not extensible.

This PR adds a new `blockEditor` action for registering additional media categories. The used setting is preserved as a private one,  so extenders can only add new categories without having any control over the core media categories.



## Tasks

- [x] Add needed type definitions to docs. 

## Testing Instructions
Run the below function in dev tools console and observe the new category in inserter's media tab. The example uses the Openverse category object and just updates the names used.
```
wp.data.dispatch( 'core/block-editor' ).registerInserterMediaCategory( {
	name: 'my-new-category',
	labels: {
		name: 'My new category',
		search_items: 'Search Openverse',
	},
	mediaType: 'image',
	async fetch( query = {} ) {
		const defaultArgs = {
			mature: false,
			excluded_source: 'flickr,inaturalist,wikimedia',
			license: 'pdm,cc0',
		};
		const finalQuery = { ...query, ...defaultArgs };
		// Sometimes you might need to map the supported request params according to `InserterMediaRequest`.
		// interface. In this example the `search` query param is named `q`.
		const mapFromInserterMediaRequest = {
			per_page: 'page_size',
			search: 'q',
		};
		const url = new URL( 'https://api.openverse.engineering/v1/images/' );
		Object.entries( finalQuery ).forEach( ( [ key, value ] ) => {
			const queryKey = mapFromInserterMediaRequest[ key ] || key;
			url.searchParams.set( queryKey, value );
		} );
		const response = await window.fetch( url, {
			headers: {
				'User-Agent': 'WordPress/inserter-media-fetch',
			},
		} );
		const jsonResponse = await response.json();
		const results = jsonResponse.results;
		return results.map( ( result ) => ( {
			...result,
			// If your response result includes an `id` prop that you want to access later, it should
			// be mapped to `InserterMediaItem`'s `sourceId` prop. This can be useful if you provide
			// a report URL getter.
			// Additionally you should always clear the `id` value of your response results because
			// it is used to identify WordPress media items.
			sourceId: result.id,
			id: undefined,
			caption: result.caption,
			previewUrl: result.thumbnail,
		} ) );
	},
	getReportUrl: ( { sourceId } ) =>
		`https://wordpress.org/openverse/image/${ sourceId }/report/`,
	isExternalResource: true,
} );
```
